src/main/java/com/offz/spigot/staminaclimb/stamina/StaminaTask.kt

```diff
@@ -6,53 +6,107 @@ import com.offz.spigot.staminaclimb.climbing.ClimbBehaviour
 
-            //regenerate stamina for BossBar
-            if (!uuid.isClimbing)
-                bar.progress = (bar.progress +
-                        //TODO this is determined by client and easily spoofable
-                        if (player.isOnGround)
-                            StaminaConfig.data.staminaRegen
-                        else StaminaConfig.data.staminaRegenInAir
-                        ).coerceAtMost(1.0)
+            if (!uuid.isClimbing) { //Regenerate stamina
+                if(player.isOnGround)
+                    StaminaBar.addProgressWithDeltaTime(StaminaConfig.data.staminaRegen
+                        .coerceAtMost(StaminaConfig.data.barMax), uuid)
+                else 
+                    StaminaBar.addProgressWithDeltaTime(StaminaConfig.data.staminaRegenInAir
+                        .coerceAtMost(StaminaConfig.data.barMax), uuid)
+            } else { //Lose stamina
+                if(isPlayerMoving(player)) {
+                    StaminaBar.removeProgressWithDeltaTime(StaminaConfig.data.staminaRemoveWhileMoving * currentWallMultiplier, uuid)
+                }
+                else {
+                    StaminaBar.removeProgressWithDeltaTime(StaminaConfig.data.staminaRemovePerTick * currentWallMultiplier, uuid)
+                }
+            }
 
-            if (progress <= StaminaConfig.data.barRed) { //Changing bar colors and effects on player depending on its progress
+            if (progress <= StaminaConfig.data.barMin) { //Changing bar colors and effects on player depending on its progress
                 bar.color = BarColor.RED
                 bar.setTitle("&c&lStamina".color()) //Make Stamina title red
-                if (uuid.isClimbing) player.stopClimbing()
-
-                uuid.canClimb = false //If player reaches red zone, they can't climb until they get back in green zone
-                player.addPotionEffect(PotionEffect(PotionEffectType.SLOW, 110, 2, false, false))
-                player.addPotionEffect(PotionEffect(PotionEffectType.WEAKNESS, 110, 2, false, false))
-            } else if (progress < 1 && !uuid.canClimb) {
+                if (uuid.isClimbing) {
+                    player.stopClimbing()
+                    uuid.canClimb = false //If player reaches red zone, they can't climb until they get back in green zone
+                    player.addPotionEffect(PotionEffect(PotionEffectType.SLOW, 110, 2, false, false))
+                    player.addPotionEffect(PotionEffect(PotionEffectType.WEAKNESS, 110, 2, false, false))
+                }
+            } else if (progress < StaminaConfig.data.barMax && !uuid.canClimb) {
                 bar.color = BarColor.RED //Keep Stamina Bar red even in yellow zone while it's regenerating
+            } else if (uuid.isClimbing && progress < StaminaConfig.data.barFlashPoint / 2.0) {
+                ticksSinceLastColorSwitch++
```

Comment:
Would it be easier to take the mod of the current value to some decimal places and somehow use that to flash?